### DISCUSSION TITLE
Turn off 'theorems' Lua filter when exporting to PDF

### DISCUSSION THREAD

**Discussion by knightgu** [quarto-dev/quarto-cli#4198] [Q&A]
I am using Quarto to write homework assignments with the LaTeX `exam` document class. The `exam` class has already defined a solution environment which can be used to show or hide the solutions with the class option `answers`. The problem I am running into is that when I export the document to PDF, Quarto always applies the Lua filter `theorems.lua` located in `src/resources/filters/crossref/theorems.lua` which adds a line `\newtheorem*{solution}{Solution}` to the exported LaTeX document and which causes LaTeX compilation error for duplicated definition. Moreover, in the exported LaTeX document, I see nested solution environments generated from the closure `::: {.solution} ... :::` (possibly due to both `theorems` and `latex-environment` filters being applied). I am wondering how I can turn off the `theorems` filter. Here is a minimal non-working example.

```
---
title: "Homework One"
format:
  pdf:
    documentclass: exam
    classoption: [12pt]
    pdf-engine: pdflatex
    filters:
      - latex-environment.lua
    environments: [questions, parts, solution]
    commands: [question, part]
  latex:
    documentclass: exam
    classoption: [12pt]
    pdf-engine: pdflatex
    filters:
      - latex-environment.lua
    environments: [questions, parts, solution]
    commands: [question, part]
---

::: {.questions}

[]{.question} Question 1

:::: {.parts}

[]{.part} 2.22 (p. 30)

::::: {.solution}
This is a very tricky problem.
:::::

::::

:::
```

**Comment by cscheid** [comment]
We don't generally support arbitrary `documentclass`es because they're not typically compatible with each other.

Some possible approaches, from most general to least:

- create a custom quarto format based on `exam` but changing the actual environments being defined in the template
- use a separate Lua filter (since you already have one) to fix the nested div issue. This is likely going to have to be a `post` filter, so your invocation will be something like:
   - ```yml
      filters:
        - latex_environment.lua
        - quarto # this is where quarto's theorems processing happens
        - fixup_nested_div.lua
      ```
- use pandoc raw blocks to undefine/redefine the latex environment as needed at the beginning of your document.

**Reply by knightgu** [reply]
Thank you for the suggestions. I totally understand the difficulty to support all the `documentclass`es in LaTeX. It is also not the main focus of Quarto. My workaround is similar to your first suggested approach. Basically, I renamed the `solution` environment of the `exam` class in the preamble which solved the nested div issue. However, the `\newtheorem*{solution}{Solution}` line is still being added during exporting. My current approach removes this line from the exported LaTeX using `sed` and then compiles the edited LaTeX to PDF. This works (and could probably also be achieved via an ad-hoc Lua filter as you suggested) but I still feel it better to probably be able to turn off the `theorems` Lua filter directly in the YAML. This could open up many possibilities.

**Reply by cderv** [reply]
@cscheid this is something we needed to handle in **bookdown** for some specific format that defines their own environment. I believe we will need that to support more Journal format that will have this constraint. 

What we did in **bookdown** was having a flag to prevent the insertion of the definition. It was the user responsibility to then define the environment in used if not done already by a documentclass or other tool. I feel we could at least do that.  

**bookdown** has also a detection mechanism by parsing the tex file to detect if a user has already manually define some environment. In that case not inclusion is done. Just another example. 

IMO we should consider

**Comment by fabbiocrux** [comment]
Hi @knightgu 
Thanks for posting this problem. I am discovering `exam` of latex using with quarto. However, reading your answer, I don't arrive to understand your workaround. 
Could you share your workaround solution please.
Thank you 🙏
